.github/workflows/scripts/build.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/python3 -u
+
+# -- prioritized --
+import sys
+import os.path
+sys.path.insert(0, os.path.join(os.path.dirname(__file__)))
+
+import common
+
+# -- stdlib --
+import glob
+import os
+import platform
+
+# -- third party --
+# -- own --
+from common.dep import download_dep
+from common.misc import get_cache_home, banner
+from common.python import setup_python
+from common.sccache import setup_sccache
+from common.tinysh import git, sccache, sh, sudo
+
+
+# -- code --
+env = os.environ
+
+
+@banner('Setup LLVM')
+def setup_llvm():
+    u = platform.uname()
+    if u.system == 'Linux':
+        if 'AMDGPU_TEST' in env:
+            # FIXME: AMDGPU bots are currently maintained separately,
+            #        we should unify them with the rest of the bots.
+            lnsf = sudo(sh.ln.bake('-sf'))
+            lnsf('/usr/bin/clang++-10', '/usr/bin/clang++')
+            lnsf('/usr/bin/clang-10',   '/usr/bin/clang')
+            lnsf('/usr/bin/ld.lld-10',  '/usr/bin/ld.lld')
+            env['LLVM_DIR'] = '/taichi-llvm-15'
+            return
+        elif common.is_manylinux2014():
+            # FIXME: prebuilt llvm15 on ubuntu didn't work on manylinux2014 image of centos. Once that's fixed, remove this hack.
+            out = get_cache_home() / 'llvm15-manylinux2014'
+            url = 'https://github.com/ailzhang/torchhub_example/releases/download/0.3/taichi-llvm-15-linux.zip'
+        else:
+            out = get_cache_home() / 'llvm15'
+            url = 'https://github.com/taichi-dev/taichi_assets/releases/download/llvm15/taichi-llvm-15-linux.zip'
+    elif (u.system, u.machine) == ('Darwin', 'arm64'):
+        out = get_cache_home() / 'llvm15-m1'
+        url = 'https://github.com/taichi-dev/taichi_assets/releases/download/llvm15/taichi-llvm-15-m1.zip'
+    elif (u.system, u.machine) == ('Darwin', 'x86_64'):
+        out = get_cache_home() / 'llvm15-mac'
+        url = 'https://github.com/taichi-dev/taichi_assets/releases/download/llvm15/llvm-15-mac10.15.zip'
+    else:
+        raise RuntimeError(f'Unsupported platform: {u.system} {u.machine}')
+
+    download_dep(url, out, strip=1)
+    env['LLVM_DIR'] = str(out)
```

Comment:
shall we just return LLVM_DIR to the caller instead of changing the environment variable directly? ;) 
In that case we can easily aggregate all the changes to environment variables in one function and it'll be easier for debugging I believe.

Reply:
I'll fix this particular site but leave others alone.

After a casual think I found it troublesome to strictly aggregate all env variables and change them at once:

1. `PATH` change in `setup_python` should immediately take effect or it would be a big surprise that the returned `python` and `pip` command couldn't work before env variable changes are applied.
2. `TAICHI_CMAKE_ARGS` is a hot spot and have been changed at multiple sites, aggregating them would require a not-so-trivial 'EnvVariableManager' to do this, that's overkill.

Debugging is a valid concern though, I'll just print the change instead of trying to 'manage' them.

Reply:
Hmm, I've changed my mind :|

After trying I found the code have a big smell of abstraction leak:

```python
def main() -> None:
    llvm_dir = setup_llvm()
    env['LLVM_DIR'] = str(llvm_dir)
    sccache = setup_sccache()
    python, pip = setup_python(env['PY'])
    build_wheel(python, pip)
    sccache('-s')
```

Reading code above, I think manipulating `LLVM_DIR` is definitely a wanted side-effect of `setup_llvm` here, at least such detail shouldn't appear in the `main` function.

Reply:
Aha I see your concern, that's totally valid. One potential solution I have in mind is passing a dict of envs into the function instead of modifying the global process's environ on the fly. `PATH` is a valid exception but other envs like `LLVM_DIR/TAICHI_CAMKE_ARGS` can be set on the subprocess (if applicable) which runs `setup.py` only.  That way we don't have to pollute the process environ just in case we start other subprocesses which may inherit the environ from it, wdyt? 
Note this isn't a blocker for this PR to merge so I'm happy to take followup PRs for it, but just want to start the discussion here to make sure the ci infra is robust ;)

Reply:
> ... `LLVM_DIR/TAICHI_CAMKE_ARGS` can be set on the subprocess (if applicable) which runs `setup.py` only. That way we don't have to pollute the process environ just in case we start other subprocesses which may inherit the environ from it, wdyt? 

Agree, but I don't think it worth the extra complexity and effort.  A simple try results pseudo-code below...

```python
def main() -> None:
    cmake_args = {}
    buildenv = {}
    cmake_args.update(split_cmake_args(env.get('TAICHI_CMAKE_ARGS', '')))
    llvm_dir, llvm_cmake_args = setup_llvm()
    cmake_args.update(llvm_cmake_args)
    buildenv['LLVM_DIR'] = llvm_dir
    sccache, sccache_cmake_args = setup_sccache()
    cmake_args.update(sccache_cmake_args)
    python, pip = setup_python(env['PY'])
    buildenv['TAICHI_CMAKE_ARGS'] = join_cmake_args(cmake_args)
    build_wheel(python, pip, buildenv)
    sccache('-s')

    distfiles = glob.glob('dist/*.whl')
    if len(distfiles) != 1:
        raise RuntimeError(f'Failed to produce exactly one wheel file: {distfiles}')
```

... which has significant more noises compared to current version.

Reply:
In bash environment vars always pollute all subprocesses but I have yet run into an issue because of this. Bash shell script writers understand this so they would careful design environment variable names.

Reply:
If env variable pollution is really a concern, I would like to use a whitelist mechanism(like what sudo does):

```python
with passthrough_envs(['TAICHI_CMAKE_ARGS', 'LLVM_DIR', 'SCCACHE_DIR']):
    python('setup.py', 'bdist_wheel')
```